[DX-1052, TT-11635] Improved body transform (#4106)
TykTechnologies/tyk-docs: tyk-docs/content/transform-traffic/request-body.md

---
title: Request Body Transformation
date: 2017-03-23T17:57:36Z
description: "How to transform the body for an API request"
tags: ["request transform", "body transform", "transform", "middleware", "per-endpoint"]
aliases:
  - /advanced-configuration/transform-traffic/request-body/
---

Tyk enables you to modify the payload of API requests before they are proxied to the upstream. This makes it easy to transform between payload data formats or to expose legacy APIs using newer schema models without having to change any client implementations. This middleware is only applicable to HTTP methods that can support a request body (i.e. PUT, POST or PATCH).

With the body transform middleware you can modify XML or JSON formatted payloads to ensure that the response contains the information required by your upstream service. You can enrich the request by adding contextual data that is held by Tyk but not included in the original request from the client.

> This middleware changes only the payload and not the headers. You can, however, combine this with the [Request Header Transform]({{< ref "transform-traffic/request-headers" >}}) to apply more complex transformation to requests.

Reviewer: ```suggestion
This middleware changes only the payload and not the headers. You can, however, combine this with the [Request Header Transform]({{< ref "transform-traffic/request-headers" >}}) middleare to apply more complex transformation to requests.
```

> There is a closely related [Response Body Transform]({{< ref "advanced-configuration/transform-traffic/response-body" >}}) middleware that provides the same functionality on the response from the upstream prior to it being returned to the client.

Reviewer: ```suggestion
There is a closely related [Response Body Transform]({{< ref "advanced-configuration/transform-traffic/response-body" >}}) middleware that provides the same functionality on the response from the upstream, prior to it being returned to the client.
```

## When to use the Request Body Transformation middleware
#### Maintaining compatibility with legacy clients

> Sometimes you might have a legacy API and need to migrate the transactions to a new upstream service but do not want to upgrade all the existing clients to the newer upstream API. Using request body transformation, you can convert the incoming legacy XML or JSON request structure into a newer, cleaner JSON format that your upstream services expect.

Reviewer: why transactions? In the scenario does the legacy API depend on an 3rd party upstream service whose API has been upgraded to a new version? Or is it that that the legacy API has changed, so using the GW to transform the client requests to be compatible with the new API?

```suggestion
Sometimes you might have a legacy API and need to migrate the requests to a new upstream service but do not want to upgrade all the existing clients to the newer upstream API. Using request body transformation, you can use the Gateway to convert the incoming legacy XML or JSON request structure into a newer, cleaner JSON format that your upstream services expect.
```
